design/fencing_mitigation.md

```diff
@@ -0,0 +1,36 @@
+# Problem Statement
+
+We need to ensure that only one node can be fenced at a time and that Ceph is fully recovered (has healthy state) before any fencing is initiated. The available pattern for limiting fencing is the MachineDisruptionBudget which allows us to specify maxUnavailable. However, this won’t be sufficient to ensure that Ceph has recovered before fencing is initiated.
+
+Therefore, we will control how many nodes match the MDB by dynamically adding and removing labels as will as dynamically updating the MDB. By manipulating the MDB into a state where desiredHealthy <= currentHealthy, we can disable fencing.
+
+# Design:
+
+## node-controller
+Each `Machine` part of a `StorageCluster` will be labeled with the key `ocs.openshift.io/<cluster-name>.
+
+The node label controller will maintain a label on the `Machine` as long as there are osds scheduled there that are not down. Therefore fencing will only be allowed on nodes with no OSDs up while the cluster is not healthy.
+
+
+The contoller will know which osds are down by querying the OSD `Deployment` objects.
```

Comment:
So here is a place where we're reaching around Rook's API (the CRs) and making assumptions about how it works. This is less dangerous than directly querying ceph, but still not ideal.